client/state/data-layer/http-data.js

```diff
@@ -184,8 +199,70 @@ export const requestHttpData = ( requestId, fetchAction, { fromApi, freshness =
 	return data;
 };
 
+/**
+ * Blocks execution until requested data has been fulfilled
+ *
+ *  - May return without data if data hasn't been fulfilled or failed
+ *  - Use for SSR contexts or when we _want_ to block rendering or execution
+ *    until the requested data has been fulfilled, e.g. when URL routing
+ *    depends on some data property
+ *  - _DO NOT USE_ when normal synchronous/data interactions suffice such
+ *    as is the case in 99.999% of React component contexts
+ *
+ * @example:
+ * withData( {
+ *     geo: () => requestGeoLocation(),
+ *     splines: () => requestSplines( siteId ),
+ * } ).then( ( { geo, splines } ) => {
+ *     return ( geo.state === 'success' || splines.state === 'success' )
+ *         ? res.send( renderToStaticMarkup( <LocalSplines geo={ geo.data } splines={ splines.data } /> ) )
+ *         : res.send( renderToStaticMarkup( <UnvailableData /> ) );
+ * }
+ *
+ * @param {object} query key/value pairs of data name and request
+ * @param {int} timeout how many ms to wait until giving up on requests
+ * @return {Promise<object>} fulfilled data of request (or partial if could not fulfill)
+ */
+export const withData = ( query, { timeout = 5000 } = {} ) =>
+	new Promise( ( resolve, reject ) => {
+		let unsubscribe = () => {};
+		let timer = null;
+		const names = Object.keys( query );
+
+		const getValues = () =>
+			names.reduce(
+				( [ values, allGood ], name ) => {
+					const value = query[ name ]();
+
+					return [ { ...values, [ name ]: value }, allGood && value.state === 'success' ];
+				},
+				[ {}, true ]
+			);
+
+		const listener = () => {
+			const [ values, allGood ] = getValues();
+
+			if ( allGood ) {
+				clearTimeout( timer );
+				unsubscribe();
+				resolve( values );
+			}
+		};
+
+		timer = setTimeout( () => {
+			const [ values ] = getValues();
+
+			unsubscribe();
+			reject( values );
```

Comment:
> Shouldn't the expired timeout also set the state of the request to timeout

No because the request might still return after the timeout. It's `pending` until it succeeds or fails. A network timeout should get indicated as much in the `failure` state. Calling code that sets a timeout should be able to know that `pending` on a failure is a timeout or it shouldn't even maybe care. The point is that by the time it rejects the data is unavailable.

Reply:
I see: timeout is not some final error condition, just a request to return early even at the cost of state of some requests being still `pending`.

I'm looking forward to see how this API will be used in practice to improve the UX in a meaningful way.

Reply:
Immediately it’s going to be used to determine whether to reroute the editor page to Gutenberg or the classic editor, a place where we have to wait until we make a decision (though hopefully in this case only on the exception that we load the editor page directly as normally we should poll this right after boot and it should be available on navigation.

_something_ like this too is a follow-up for our chat one ČB and I hope can provide some use guidance as we expand SSR, but that’s a distant secondary goal here